lib/OrderedMap.fram

```diff
@@ -0,0 +1,315 @@
+
+method remove {X, self : Tree X} elem lt = delete elem self lt
+
+let insert tree elem lt = 
+  let rec ins tree =  
+    match tree with 
+    | Leaf => Node Red Leaf elem Leaf
+    | Node color a y b => 
+      if lt elem y then balance (Node color (ins a) y b)
+      else if lt y elem then balance (Node color a y (ins b))
+      else Node color a y b 
+    end
+  let y = ins tree 
+  in match y with
+  | Node _ y a b => Node Black y a b
+  | Leaf => Leaf
+  end
+
+method add {X, self : Tree X} elem lt = insert self elem lt
+
+let toList tree = 
+  let rec _toList tree acc =
+    match tree with 
+    | Leaf => acc
+    | Node _ a y b => _toList a (y :: _toList b acc)
+    end
+  in _toList tree []
+
+method toList {X , self : Tree X} = toList self
+
+let fromList {X} (lista : List X) lt = 
+  let rec _fromList lista (acc : Tree X) =
+    match lista with 
+    | x :: xs => _fromList xs (acc.add x lt)
+    | [] => acc
+    end 
+  in _fromList lista Leaf
+
+let rec foldlRB fun acc from = 
+  match from with
+  | Leaf => acc
+  | Node _ a y b => 
+  let left = foldlRB fun acc a in 
+  let middle = (fun y left) in
+  foldlRB fun middle b
+end
+
+let rec foldrRB fun acc from = 
+match from with
+  | Leaf => acc
+  | Node _ a y b => 
+  let right = foldlRB fun acc b in 
+  let middle = (fun y right) in
+  foldlRB fun middle a
+end
+
+method traversL {X, self : Tree X} fun acc = foldlRB fun acc self
+method traversR {X, self : Tree X} fun acc = foldrRB fun acc self
+
+//TODO: better merge
+let merge {X} (from : Tree X) (to : Tree X) lt = 
+  foldlRB (fn x (y : Tree X) => y.add x lt) to from
+
+let rec member elem tree lt =
+  match tree with 
+  | Leaf => False
+  | Node _ l v r =>
+    if lt elem v then member elem l lt
+    else if lt v elem then member elem r lt
+    else True
+  end
+
+
+method member {X, self : Tree X} elem lt  = member elem self lt
+
+
+let add lt tree key value = insert tree (key,value) lt
+let join lt tree1 tree2 = merge tree1 tree2 lt
+let rmVal lt key tree = match tree with 
+  | Leaf => Leaf
+  | Node _ _ (_,v) _ => delete (key,v) tree lt
+end
+
+let rec _find lt elem tree = 
+  match tree with 
+  | Leaf => None
+  | Node _ l v r => if lt elem v then _find lt elem l 
+  else if lt v elem then _find lt elem r else Some v
+end
+
+let find lt key tree = 
+  match tree with 
+  | Leaf => None
+  | Node _ l v r => let (_,a) = v in _find lt (key,a) tree
+end 
+
+let update lt tree key val = let x = rmVal lt key tree in
+add lt x key val
+
+let fst (a,b) = a
+let snd (a,b) = b
+
+data OrderedMap Key Val = Map of 
+{
+  T
+  , empty : T
+  , method add : T  -> Key -> Val -> [] T
+  , method join : T -> T -> [] T
+  , method isEmpty : T -> [] Bool
+  , method deleteElem : Key -> T -> [] T
+  , method toList : T -> [] List (Pair Key Val)
+  , method update : T -> Key -> Val -> [] T
+  , method deleteMax : T -> [] T
+  , method deleteMin : T -> [] T
+}
+
+pub let make {Key} {Val} (lt : Key -> [] Key -> [] Bool)= 
```

Comment:
I think `lt` should be a named parameter, and also it is cleaner if all named parameters are in one pair of brackets